rbd: fail fast in create volume for missmatch encryption

Madhu-1 opened the pull request:
CreateVolume will fail in the below cases
    
### create volume from the snapshot
* If the snapshot is encrypted and the requested volume  is not encrypted
* If the snapshot is not encrypted and the requested volume is encrypted
    
### create volume from volume
* If the parent volume is encrypted and requested volume is not encrypted
* If the parent volume is not encrypted and the requested volume is encrypted
    
Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

nixpanic commented:
> verifying encrypted pvc creation from unencrypted PVC works #2109

This is not expected to work. Creating an encrypted volume from an unencrypted source is not going to be supported. If this is a feature that is needed, we need to:

1. create a new, empty RBD image
2. attach the new RBD image, setup encryption with `cryptsetup` (creates a new encrypted block-device)
3. read the data from the source RBD image, and write it to the encrypted block-device on top of the destination RBD image

Madhu-1 committed:
rbd: fail fast in create volume for missmatch encryption

CreateVolume will fail in below cases

* If the snapshot is encrypted and requested volume
is not encrypted
* If the snapshot is not encrypted and requested
volume is encrypted

* If the parent volume is encrypted and requested volume
is not encrypted
* If the parent volume is not encrypted and requested
volume is encrypted

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 commented:
> > verifying encrypted pvc creation from unencrypted PVC works #2109
> 
> This is not expected to work. Creating an encrypted volume from an unencrypted source is not going to be supported. If this is a feature that is needed, we need to:
> 
> 1. create a new, empty RBD image
> 2. attach the new RBD image, setup encryption with `cryptsetup` (creates a new encrypted block-device)
> 3. read the data from the source RBD image, and write it to the encrypted block-device on top of the destination RBD image

There is no requirement yet for this one. This PR fails fast in create volume request instead of failing at mounting the PVC.

Rakshith-R commented on internal/rbd/controllerserver.go:
```diff
@@ -217,6 +217,37 @@ func validateRequestedVolumeSize(rbdVol, parentVol *rbdVolume, rbdSnap *rbdSnaps
+func checkValidEncryptionRequest(rbdVol, parentVol *rbdVolume, rbdSnap *rbdSnapshot) error {
+	// if snapshot is encrypted the request volume should also be encrypted
+	if rbdSnap != nil {
+		if rbdSnap.isEncrypted() && !rbdVol.isEncrypted() {
+			return status.Error(codes.InvalidArgument, "cannot create unencrypted volume from encrypted snapshot")
+		}
+		if !rbdSnap.isEncrypted() && rbdVol.isEncrypted() {
+			return status.Error(codes.InvalidArgument, "cannot create encrypted volume from unencrypted snapshot")
+		}
```
```suggestion
		switch {
		case rbdSnap.isEncrypted() && !rbdVol.isEncrypted():
			return status.Error(codes.InvalidArgument, "cannot create unencrypted volume from encrypted snapshot")

		case !rbdSnap.isEncrypted() && rbdVol.isEncrypted():
			return status.Error(codes.InvalidArgument, "cannot create encrypted volume from unencrypted snapshot")
		}
```

Madhu-1 replied:
I don't have any preference. will wait for others' suggestions.

nixpanic replied:
I think I prefer the switch/case, but I do not care *that* much either.

humblec replied:
Yeah. `switch` will be cleaner I believe.

Rakshith-R commented on internal/rbd/controllerserver.go:
```diff
@@ -217,6 +217,37 @@ func validateRequestedVolumeSize(rbdVol, parentVol *rbdVolume, rbdSnap *rbdSnaps
+	}
+	// if parent volume is encrypted the request volume should also be encrypted
+	if parentVol != nil {
+		if parentVol.isEncrypted() && !rbdVol.isEncrypted() {
+			return status.Error(codes.InvalidArgument, "cannot create unencrypted volume from encrypted volume")
+		}
+		if !parentVol.isEncrypted() && rbdVol.isEncrypted() {
+			return status.Error(codes.InvalidArgument, "cannot create encrypted volume from unencrypted volume")
+		}
```
```suggestion
		switch {
		case parentVol.isEncrypted() && !rbdVol.isEncrypted():
			return status.Error(codes.InvalidArgument, "cannot create unencrypted volume from encrypted volume")

		case !parentVol.isEncrypted() && rbdVol.isEncrypted():
			return status.Error(codes.InvalidArgument, "cannot create encrypted volume from unencrypted volume")
		}
```

Rakshith-R reviewed (commented):
Using switch instead if statements a bit more cleaner ?

nixpanic commented on internal/rbd/controllerserver_test.go:
```diff
@@ -0,0 +1,163 @@
+/*
+Copyright 2018 The Ceph-CSI Authors.
```
date?

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -217,6 +217,37 @@ func validateRequestedVolumeSize(rbdVol, parentVol *rbdVolume, rbdSnap *rbdSnaps
+	return checkValidEncryptionRequest(rbdVol, parentVol, rbdSnap)
+}
+
+func checkValidEncryptionRequest(rbdVol, parentVol *rbdVolume, rbdSnap *rbdSnapshot) error {
```
can you use `src, dst *rbdImage` as arguments to this function?

Madhu-1 commented:
/test ci/centos/k8s-e2e-external-storage/1.20

Madhu-1 commented:
@nixpanic PTAL

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -210,6 +210,39 @@ func validateRequestedVolumeSize(rbdVol, parentVol *rbdVolume, rbdSnap *rbdSnaps
+	return checkValidEncryptionRequest(parentVol, rbdVol, rbdSnap)
+}
+
+func checkValidEncryptionRequest(src, dest *rbdVolume, rbdSnap *rbdSnapshot) error {
```
A snapshot is encrypted if its parent volume is encrypted. I do not think there is a need to pass `rbdSnap` here.

```go
func checkValidEncryptionRequest(src, dest *rbdImage) error {
```

and then call  this function only with the source (parentVol or rbdSnap) and the destination. Handling multiple cases in this function is more complex and confusing than is needed.

Madhu-1 replied:
we are passing dest and rbdSnap for two different CreateVolume cases. One is with PVC-PVC clone and Restore snapshot to a new PVC. not sure how I can manage with only src and dest here.

nixpanic replied:
Sure, but is there not always a single source and a single destination? In `checkValidCreateVolumeRequest` you can call `checkValidEncryptionRequest` depending on `parentVol` or `rbdSnap`? Use `rbdSnap` unless it is `nil`, otherwise use `parentVol`?

Madhu-1 replied:
In one or another place i need to check for nil objects. I thought doing it in checkValidEncryptionRequest and send proper detailed error message. Do you have any suggestions how the function(pseudo code) should be like?

nixpanic replied:
maybe something like this:

```go
func checkValidCreateVolumeRequest(rbdVol, parentVol *rbdVolume, rbdSnap *rbdSnapshot, cr *util.Credentials) error {
        err := validateRequestedVolumeSize(rbdVol, parentVol, rbdSnap, cr)
        if err != nil {
                return err
        }

        if rbdSnap != nil {
                err = checkValidEncryptionRequest(&rbdSnap.rbdImage, &rbdVol.rbdImage)
        } else {
                err = checkValidEncryptionRequest(&parentVol.rbdImage, &rbdVol.rbdImage)
        }

        return err
}

// checkValidEncryptionRequest verifies that both src and dest have the same
// encryption status.
func checkValidEncryptionRequest(src, dest *rbdImage) error {
        switch {
        case src.isEncrypted() && !dest.isEncrypted():
                return status.Errorf(codes.InvalidArgument, "cannot create unencrypted volume from encrypted volume")
        
        case !src.isEncrypted() && dest.isEncrypted():
                return status.Errorf(codes.InvalidArgument, "cannot create encrypted volume from unencrypted volume")
        }
        return nil
}
```

The error message should be clear in any case, the user would know if a clone or restore from snapshot is done.

Madhu-1 replied:
@nixpanic but with the above pseudo-code even in the case of Restore from snapshot it's returning a common error message which holds good for PVC-PVC clone.

nixpanic reviewed (changes requested):
Please check if you can simplify the new function. `rbdVolume` and `rbdSnapshot` are both `rbdImage`. Passing 3 arguments for a function when for each invocation only 2 are really required is not very clean.

In case I'm mistaking, a clear comment above the function would be justified :-)

Madhu-1 commented:
> assing 3 arguments for a function when for each invocation only 2 are really required is not very clean.

Not sure simplifying helps here. we are passing 3 arguments rbbVol,parentVol,snap to other function also `createBackingImage`  if really needed in will add two separate functions and also will add ` nolint:gocyclo // golangci-lint` to the CreateVolume as it increases the complexity of CreateVolume function.

Madhu-1 commented:
@nixpanic PTAL implemented the suggestion
